pex/jobs.py

```diff
@@ -0,0 +1,243 @@
+  :returns: A job handle to the spawned python process.
+  :rtype: :class:`Job`
+  """
+  if expose:
+    subprocess_env = (env or os.environ).copy()
+    # In order to expose vendored distributions with their un-vendored import paths in-tact, we
+    # need to set `__PEX_UNVENDORED__`. See: vendor.__main__.ImportRewriter._modify_import.
+    subprocess_env['__PEX_UNVENDORED__'] = '1'
+
+    pythonpath = third_party.expose(expose)
+  else:
+    subprocess_env = env
+    pythonpath = None
+
+  interpreter = interpreter or PythonInterpreter.get()
+  cmd, process = interpreter.open_process(
+    args=args,
+    pythonpath=pythonpath,
+    env=subprocess_env,
+    **subprocess_kwargs
+  )
+  return Job(command=cmd, process=process)
+
+
+class SpawnedJob(object):
+  """A handle to a spawned :class:`Job` and its associated result."""
+
+  @classmethod
+  def wait(cls, job, result):
+    """Wait for the job to complete and return a fixed result upon success.
+
+    :param job: The spawned job.
+    :type job: :class:`Job`
+    :param result: The fixed success result.
+    :return: A spawned job whose result is a side effect of the job (a written file, a populated
+             directory, etc.).
+    :rtype: :class:`SpawnedJob`
+    """
+    def wait_result_func():
+      job.wait()
+      return result
+
+    return cls(job=job, result_func=wait_result_func)
+
+  @classmethod
+  def stdout(cls, job, result_func, input=None):
+    """Wait for the job to complete and return a result derived from its stdout.
+
+    :param job: The spawned job.
+    :type job: :class:`Job`
+    :param result_func: A function taking the stdout byte string collected from the spawned job and
+                        returning the desired result.
+    :param input: Optional input stream data to pass to the process as per the
+                  `subprocess.Popen.communicate` API.
+    :return: A spawned job whose result is derived from stdout contents.
+    :rtype: :class:`SpawnedJob`
+    """
+    def stdout_result_func():
+      stdout, _ = job.communicate(input=input)
+      return result_func(stdout)
+
+    return cls(job=job, result_func=stdout_result_func)
+
+  def __init__(self, job, result_func):
+    """Not intended for direct use, see `wait` and `stdout` factories."""
+    self._job = job
+    self._result_func = result_func
+
+  def await_result(self):
+    """Waits for the spawned job to complete and returns its result."""
+    return self._result_func()
+
+  def kill(self):
+    """Terminates the spawned job if it's not already complete."""
+    self._job.kill()
+
+  def __str__(self):
+    return str(self._job)
+
+
+_CPU_COUNT = cpu_count()
+_ABSOLUTE_MAX_JOBS = _CPU_COUNT * 2
+
+
+DEFAULT_MAX_JOBS = _CPU_COUNT
+"""The default maximum number of parallel jobs PEX should use."""
+
+
+def _sanitize_max_jobs(max_jobs=None):
+  assert max_jobs is None or isinstance(max_jobs, int)
+  if max_jobs is None or max_jobs <= 0:
```

Comment:
Perhaps, but as of now pex has no warning system, only -v(vvvvvvvvv). Just below the actual number used is logged at maximally fine level (9).

Reply:
Ok, great! I generally interpret an argument named like `--max-parallel-jobs` as a suggestion to the tool instead of a hard line, so the sanitizing here sounds quite reasonable.